2.5-huffman.ipynb

```diff
@@ -1,25 +1,248 @@
 {
  "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import util\n",
+    "from test_framework.key import ECKey, ECPubKey, generate_key_pair\n",
+    "from test_framework.script import TapTree, TapLeaf, Node"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# 2.5 Huffman TapTree Constructor\n",
+    "\n",
+    "When constructing a TapTree, there is the possibility of optimizing the tree structure so that the expected witness weight when spending along the script path is minimized. The `expected witness weight` is determined by summing up all probability-adjusted witness weights of the TapScripts in the TapTree.\n",
+    "\n",
+    "For example, for a TapTree with TapLeafs A, B, and C, the expected witness weight is determined as follows:\n",
+    "\n",
+    "* **`Expected-witness-weight =`**\n",
+    "    * **`Probability-of-A * Witness-weight-A`**\n",
+    "    * **`+ Probability-of-B * Witness-weight-B`**\n",
+    "    * **`+ Probability-of-C * Witness-weight-C`**\n",
+    "\n",
+    "In order to minimize the expected witness weight for the script path, we can try to reduce the size of the required `inclusion proof` for those TapScripts which have a higher probability weighting by placing these closer to the root of the tree."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Note:** There may be commonly used Taproot output templates with known TapScripts and respective probabilities. Spending a TapScript will reveal both the leaf depth and TapScript, and may allow an observer to match this information against known standard output types. In this case, there exists a trade-off between spending cost and privacy."
```

Comment:
I'd suggest dropping this. It's confusing and distracts from the main point of the chapter.

Reply:
Ok. Removed.